### PR TITLE
Partitions

### PR DESCRIPTION
**Pull request by daniel-davidd**

### CODE REVIEW

File: memphis/memphis.py [hunk 1]
```diff
@@ -599,15 +612,21 @@ async def consumer(
...
+            creation_res = json.loads(creation_res)
 
-            if err_msg != "":
-                raise MemphisError(err_msg)
 
+            if creation_res["error"] != "":
+                raise MemphisError(creation_res["error"])
             internal_station_name = get_internal_name(station_name)
+
+            if not creation_res["partitions_update"]["partitions_list"] is None:
```
Comment (rnowling-memphis) [inline memphis/memphis.py:626]: please move the "not" before the "None":

`if creation_res["partitions_update"]["partitions_list"] is not None:`

File: memphis/partition_generator.py [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+class PartitionGenerator:
+    def __init__(self, partitions_list):
+        self.partitions_list = partitions_list
+        self.current = 0
+        self.length = len(partitions_list)
+
+    def next(self):
+        partition_to_return = self.partitions_list[self.current]
+        self.current = (self.current + 1) % self.length
+        return partition_to_return
```
Comment (rnowling-memphis) [inline memphis/partition_generator.py:12]: Missing end of file marker

File: memphis/partition_generator.py [hunk 3]
```diff
@@ -0,0 +1,12 @@
+
```
Comment (rnowling-memphis) [inline memphis/partition_generator.py:1]: Remove blank lines at top

File: memphis/partition_generator.py [hunk 4]
```diff
@@ -0,0 +1,12 @@
...
+
+class PartitionGenerator:
+    def __init__(self, partitions_list):
+        self.partitions_list = partitions_list
+        self.current = 0
+        self.length = len(partitions_list)
+
+    def next(self):
```
Comment (rnowling-memphis) [inline memphis/partition_generator.py:9]: If you rename this to `__next__`, you can then use the more Pythonic approach of working with iterators: `next(iter)`

File: memphis/partition_generator.py [hunk 5]
```diff
@@ -0,0 +1,12 @@
+
+
+class PartitionGenerator:
```
Comment (rnowling-memphis) [inline memphis/partition_generator.py:3]: Can you add some documentation indicating what this class does and how it should be used?

File: memphis/producer.py [hunk 6]
```diff
@@ -232,6 +235,11 @@ async def produce(
             else:
                 headers = memphis_headers
 
+            if not self.internal_station_name in self.connection.partition_producers_updates_data:
```
Comment (rnowling-memphis) [inline memphis/producer.py:238]: Change to `if self.internal_station_name not in self.connection.partition_producers_updates_data`

File: memphis/memphis.py [hunk 7]
```diff
@@ -420,6 +427,12 @@ async def producer(
                 raise MemphisError(create_res["error"])
 
             internal_station_name = get_internal_name(station_name)
+
+            if not create_res["partitions_update"]["partitions_list"] is None:
```
Comment (rnowling-memphis) [inline memphis/memphis.py:431]: Please change to `if create_res["partitions_update"]["partitions_list"] is not None`

File: memphis/consumer.py [hunk 8]
```diff
@@ -91,16 +97,29 @@ async def main():
         self.t_consume = asyncio.create_task(self.__consume(callback))
 
     async def __consume(self, callback):
-        subject = get_internal_name(self.station_name)
-        consumer_group = get_internal_name(self.consumer_group)
-        self.psub = await self.connection.broker_connection.pull_subscribe(
-            subject + ".final", durable=consumer_group
-        )
+        if not self.inner_station_name in self.connection.partition_consumers_updates_data:
```
Comment (rnowling-memphis) [inline memphis/consumer.py:100]: Please change to `if self.inner_station_name not in self.connection.partition_consumers_updates_data`

File: memphis/consumer.py [hunk 9]
```diff
@@ -229,11 +248,22 @@ async def main(host, username, password, station):
...
         while True:
             try:
+                
                 await asyncio.sleep(self.ping_consumer_interval_ms / 1000)
+                station_inner = get_internal_name(self.station_name)
                 consumer_group = get_internal_name(self.consumer_group)
-                await self.connection.broker_connection.consumer_info(
-                    self.station_name, consumer_group, timeout=30
-                )
+                if not self.inner_station_name in self.connection.partition_consumers_updates_data:
```
Comment (rnowling-memphis) [inline memphis/consumer.py:255]: Please change to `if self.inner_station_name not in self.connection.partition_consumers_updates_data`

**Review by rnowling-memphis — CHANGES_REQUESTED** [review]
Please fix any associated Pylint issues.

**Review by idanasulinmemphis — APPROVED** [review]

**Pull request merged**